src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionSelector.cs

```diff
@@ -234,5 +285,217 @@ protected virtual IReadOnlyList<ActionDescriptor> SelectBestActions(IReadOnlyLis
+                        // with this action (in order) so we can store them in our maps.
+                        var values = new string[RouteKeys.Length];
+                        for (var j = 0; j < RouteKeys.Length; j++)
+                        {
+                            action.RouteValues.TryGetValue(RouteKeys[j], out values[j]);
+                        }
+
+                        actionsAndKeys[count++] = (action, values);
+
+                        List<ActionDescriptor> entries;
+                        if (!OrdinalIgnoreCaseEntries.TryGetValue(values, out entries))
+                        {
+                            entries = new List<ActionDescriptor>();
+                            OrdinalIgnoreCaseEntries.Add(values, entries);
+                        }
+
+                        entries.Add(action);
+                    }
+                }
+
+                // At this point `OrdinalIgnoreCaseEntries[key].Value` contains all of the actions in the same
+                // equivalence class - meaning, all conventionally routed actions for which the route values are equal
+                // ignoring case.
+                //
+                // Now we want to iterate the actions **again** and for each unique casing, map this back to the full
+                // set for that equivalence class.
+                for (var i = 0; i < actionsAndKeys.Length; i++)
+                {
+                    var actionAndKeys = actionsAndKeys[i];
+                    if (!OrdinalEntries.ContainsKey(actionAndKeys.keys))
+                    {
+                        OrdinalEntries[actionAndKeys.keys] = OrdinalIgnoreCaseEntries[actionAndKeys.keys];
+                    }
+                }
+            }
+
+            public int Version { get; set; }
+
+            public string[] RouteKeys { get; set; }
+
+            public Dictionary<string[], List<ActionDescriptor>> OrdinalEntries { get; set; }
+
+            public Dictionary<string[], List<ActionDescriptor>> OrdinalIgnoreCaseEntries { get; set; }
+        }
+
+        private class OrdinalEqualityComparer : IEqualityComparer<string[]>
+        {
+            public static readonly IEqualityComparer<string[]> Instance = new OrdinalEqualityComparer();
+
+            public bool Equals(string[] x, string[] y)
+            {
+                if (object.ReferenceEquals(x, y))
+                {
+                    return true;
+                }
+
+                if (x == null ^ y == null)
+                {
+                    return false;
+                }
+
+                if (x.Length != y.Length)
+                {
+                    return false;
+                }
+
+                for (var i = 0; i < x.Length; i++)
+                {
+                    if (string.IsNullOrEmpty(x[i]) && string.IsNullOrEmpty(y[i]))
+                    {
+                        continue;
```

Comment:
This file goes to some lengths to store `string.Empty` and `null` rather than to map one to the other. Why bother if they're considered equivalent everywhere else?